p2p/src/sync/peer.rs

```diff
@@ -317,6 +320,26 @@ where
     async fn handle_header_list(&mut self, headers: Vec<SignedBlockHeader>) -> Result<()> {
         log::debug!("Headers list from peer {}", self.id());
 
+        if let Some(last_header) = headers.last() {
+            // Delays the processing of a new block until it can be accepted by the chainstate (but not more than `max_clock_diff`).
+            // This is needed to allow the local or remote node to have slightly inaccurate clocks.
+            // Without it, even a 1 second difference can break block synchronization
+            // because one side may see the new block as invalid.
+            let block_timestamp = last_header.timestamp().as_duration_since_epoch();
+            let max_block_timestamp =
+                self.time_getter.get_time() + *self.chain_config.max_future_block_time_offset();
```

Comment:
There's also the tolerance from the chainstate... blocks are allowed to be off by that much (which is 120 seconds now).

Please use checked_add when adding that into account instead of direct addition that can panic.

Reply:
> There's also the tolerance from the chainstate... blocks are allowed to be off by that much (which is 120 seconds now).

Yes, sure, but I already use it here (`max_future_block_time_offset` equals to 120 seconds).

> Please use checked_add when adding that into account instead of direct addition that can panic.

I think it's OK, both `time_getter` and `max_future_block_time_offset` can be trusted.